### PR TITLE
Reworked Stale Developers Logic

### PR DESCRIPTION
**Pull request by anoosh96**
this will address #840

- [x] My code contains tests covering the code I modified
- [x] I linted and tested the project with `bin/check`
- [ ] I added significant changes and product updates to the [changelog](CHANGELOG.md)

### CODE REVIEW

**Review by joemasilotti — COMMENTED** [review]
Do we need to keep track of if a developer was notified anymore?

With the new logic, we will be marking stale profiles as not interested. If someone isn't interested I don't care if their profile hasn't been updated in a month - and I don't think we need to notify them about it.

I think the new logic can check if the developer is actively (looking or open to offers) _and_ hasn't been updated in more than 1 month. If so, mark them as not interested and send the email.

The next time the query is run they won't be included because they are no longer actively looking or open to offers.

Does that make sense? If so, would you like to apply the necessary changes?

**Edit**: We can use `Developer.actively_looking_or_open` to include `nil`, too. And lets change the threshold from one month to three.

**Comment by anoosh96** [comment]
> Do we need to keep track of if a developer was notified anymore?
> 
> With the new logic, we will be marking stale profiles as not interested. If someone isn't interested I don't care if their profile hasn't been updated in a month - and I don't think we need to notify them about it.
> 
> I think the new logic can check if the developer is actively (looking or open to offers) _and_ hasn't been updated in more than 1 month. If so, mark them as not interested and send the email.
> 
> The next time the query is run they won't be included because they are no longer actively looking or open to offers.
> 
> Does that make sense? If so, would you like to apply the necessary changes?
> 
> **Edit**: We can use `Developer.actively_looking_or_open` to include `nil`, too. And lets change the threshold from one month to three.

@joemasilotti  that depends more on the requirements if we want to keep reminding the developer that their profile has been inactivated or just notify them once. I was thinking about, what if the developer misses the email?. If we don't want to keep notifying them then it totally makes sense. I will make these changes

**Comment by joemasilotti** [comment]
I’m fine with only notifying them once. If they miss the email, that’s in them in my opinion. 

And because we are sending it via a notification we will always have a record that we sent it. So we can update logic accordingly in the future.

**Comment by anoosh96** [comment]
> I’m fine with only notifying them once. If they miss the email, that’s in them in my opinion.
> 
> And because we are sending it via a notification we will always have a record that we sent it. So we can update logic accordingly in the future.

that makes sense, thanks for the clarification.

**Comment by anoosh96** [comment]
Hey @joemasilotti, I have made the requested changes.

**Review by joemasilotti — CHANGES_REQUESTED** [review]
This is looking good! I made a few requests and pushed some copy updates.

File: app/models/concerns/developers/notifications.rb [hunk 1]
```diff
@@ -20,8 +20,8 @@ def invisiblize_and_notify!
       send_invisiblize_notification
     end
 
-    def notify_as_stale
-      send_stale_profile_notification
+    def mark_as_stale_and_notify
+      send_stale_profile_notification if mark_stale
```
[the remark this answers is not in the document]
Comment (anoosh96) [reply app/models/concerns/developers/notifications.rb:24]: I think update_column can return false as well if something goes wrong with updating the column (there's a possibility), ref:
[](https://apidock.com/rails/ActiveRecord/Persistence/update_columns)

File: app/queries/stale_developers_query.rb [hunk 2]
```diff
@@ -1,43 +1,8 @@
 class StaleDevelopersQuery
-  EARLIEST_TIME = 30.days.ago.beginning_of_day
+  EARLIEST_TIME = 3.months.ago.beginning_of_day
 
-  def stale_and_not_recently_notified
+  def stale_developers
```
[the remark this answers is not in the document]
Comment (anoosh96) [reply app/queries/stale_developers_query.rb:4]: yes, I better move this into Custodian, as its only being used in one place.

File: app/models/developers/custodian.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+module Developers
+  class Custodian
+    def self.clean_stale_profiles
+      stale_developers = StaleDevelopersQuery.new.stale_developers
+      stale_developers.each(&:mark_as_stale_and_notify)
+      stale_developers.profile_reminder_notifications
```
[the remark this answers is not in the document]
Comment (anoosh96) [reply app/models/developers/custodian.rb:6]: it returns the developers that were marked stale and notified, and we are logging that number in the rake task, we can move the logging to Custodian class if required.

**Comment by anoosh96** [comment]
> This is looking good! I made a few requests and pushed some copy updates.

Hey @joemasilotti Thanks for the commits, I really missed out on those little things, also I have pushed some changes regarding the comments, looking forward to your feedback.

File: app/models/concerns/developers/notifications.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (joemasilotti) [reply app/models/concerns/developers/notifications.rb:24]: Got it, thanks!

File: app/models/developers/custodian.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (joemasilotti) [reply app/models/developers/custodian.rb]: Ah, right. Thanks for clarifying.

**Review by joemasilotti — APPROVED** [review]
This looks great, thanks @anoosh96!

I'm going to wrap it in a feature flag as I work on one other coordinated issue. But please [email me](mailto:joe@masilotti.com) how you'd like to receive your bounty (Venmo, PayPal, Zelle, or Wise).

**Comment by anoosh96** [comment]
Hey @joemasilotti, I am actually located in Pakistan, so these won't work for me, can I send you my IBAN, I mostly recieve international payments through my IBAN.

**Comment by joemasilotti** [comment]
I'm not sure what IBAN is but I bet we can figure something out! Email me so we don't end up sending back information in public. :)

**Pull request closed without merging**